lib/ansible/modules/cloud/amazon/s3_bucket.py

```diff
@@ -156,172 +130,185 @@ def _create_or_update_bucket(connection, module, location):
+                except (BotoCoreError, ClientError) as e:
+                    module.fail_json_aws(e, msg="Failed to delete bucket tags")
+            current_tags_dict = tags
+            changed = True
+
+    module.exit_json(changed=changed, name=name, versioning=versioning_return_value,
+                     requester_pays=requester_pays, policy=current_policy, tags=current_tags_dict)
+
 
-    module.exit_json(changed=changed, name=bucket.name, versioning=versioning_status,
-                     requester_pays=requester_pays_status, policy=current_policy, tags=current_tags_dict)
+def paginated_list(s3_client, **pagination_params):
+    pg = s3_client.get_paginator('list_objects_v2')
+    for page in pg.paginate(**pagination_params):
+        yield [data['Key'] for data in page.get('Contents', [])]
 
 
-def _destroy_bucket(connection, module):
+def destroy_bucket(s3_client, module):
 
     force = module.params.get("force")
     name = module.params.get("name")
     changed = False
-
     try:
-        bucket = connection.get_bucket(name)
-    except S3ResponseError as e:
-        if e.error_code != "NoSuchBucket":
-            module.fail_json(msg=e.message)
-        else:
+        s3_client.head_bucket(Bucket=name)
+    except EndpointConnectionError as e:
+        module.fail_json_aws(e, msg="Invalid endpoint provided: %s" % to_text(e))
+    except ClientError as e:
+        # If a client error is thrown, then check that it was a 404 error.
+        # If it was a 404 error, then the bucket does not exist.
+        error_code = int(e.response['Error']['Code'])
+        if error_code == 404:
             # Bucket already absent
             module.exit_json(changed=changed)
+        else:
+            module.fail_json_aws(e, msg="Failed to check bucket presence")
+    except BotoCoreError as e:
+        module.fail_json_aws(e, msg="Failed to check bucket presence")
 
     if force:
+        # if there are contents then we need to delete them before we can delete the bucket
         try:
-            # Empty the bucket
-            for key in bucket.list():
-                key.delete()
-
-        except BotoServerError as e:
-            module.fail_json(msg=e.message)
+            for keys in paginated_list(s3_client, Bucket=name):
+                formatted_keys = [{'Key': key} for key in keys]
+                if formatted_keys:
+                    s3_client.delete_objects(Bucket=name, Delete={'Objects': formatted_keys})
+            changed = True
+        except (BotoCoreError, ClientError) as e:
+            module.fail_json_aws(e, msg="Failed while deleting bucket")
 
     try:
-        bucket = connection.delete_bucket(name)
-        changed = True
-    except S3ResponseError as e:
-        module.fail_json(msg=e.message)
+        s3_client.delete_bucket(Bucket=name)
+    except (BotoCoreError, ClientError) as e:
+        module.fail_json_aws(e, msg="Failed to delete bucket")
```

Comment:
I'm getting sporadic errors in the tests with this:
```
fatal: [localhost]: FAILED! => {
    "changed": false,
    "error": {
        "bucket_name": "ansible-test-shertel-osx-73152613.testbucket.ansible",
        "code": "NoSuchBucket",
        "message": "The specified bucket does not exist"
    },
    "invocation": {
        "module_args": {
            "aws_access_key": "redacted",
            "aws_secret_key": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER",
            "ceph": false,
            "ec2_url": null,
            "force": false,
            "name": "ansible-test-shertel-osx-73152613.testbucket.ansible",
            "policy": null,
            "profile": null,
            "region": "us-east-1",
            "requester_pays": false,
            "s3_url": null,
            "security_token": "",
            "state": "absent",
            "tags": null,
            "validate_certs": true,
            "versioning": null
        }
    },
    "msg": "Failed to delete bucket: An error occurred (NoSuchBucket) when calling the DeleteBucket operation: The specified bucket does not exist",
    "response_metadata": {
        "host_id": "kfrPIUSyMWBlirnpkbgMHiUbYgrq1LyYF50BVqdwKpnPJ4UuR3CB54TgbQ+cU+s7+Hlq6iTWdRM=",
        "http_headers": {
            "content-type": "application/xml",
            "date": "Fri, 23 Mar 2018 13:43:56 GMT",
            "server": "AmazonS3",
            "transfer-encoding": "chunked",
            "x-amz-id-2": "kfrPIUSyMWBlirnpkbgMHiUbYgrq1LyYF50BVqdwKpnPJ4UuR3CB54TgbQ+cU+s7+Hlq6iTWdRM=",
            "x-amz-request-id": "4E6E0CBF718620B7"
        },
        "http_status_code": 404,
        "request_id": "4E6E0CBF718620B7",
        "retry_attempts": 0
    }
}
```
I think the issue is it may still be creating or is in the process of being deleted (I think both of those because I ran the tests twice and it failed twice -different tests- with the same error, but I only had one bucket to manually delete). Using the S3 BucketExists and BucketNotExists waiters would probably fix this.

I'm also seeing a sporadic error with idempotence for creation, which I think would be fixed with the BucketExists waiter:
```
TASK [s3_bucket : Try to update the same bucket with the same values] **********************************************************
task path: /Users/shertel/Workspace/ansible/test/integration/targets/s3_bucket/tasks/main.yml:30
Using module file /Users/shertel/Workspace/ansible/lib/ansible/modules/cloud/amazon/s3_bucket.py
<127.0.0.1> ESTABLISH LOCAL CONNECTION FOR USER: shertel
<127.0.0.1> EXEC /bin/sh -c 'echo ~ && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /Users/shertel/.ansible/tmp/ansible-tmp-1521813525.36-213064420310223 `" && echo ansible-tmp-1521813525.36-213064420310223="` echo /Users/shertel/.ansible/tmp/ansible-tmp-1521813525.36-213064420310223 `" ) && sleep 0'
<127.0.0.1> PUT /Users/shertel/.ansible/tmp/ansible-local-97538M9gW9j/tmpjcF4y2 TO /Users/shertel/.ansible/tmp/ansible-tmp-1521813525.36-213064420310223/s3_bucket.py
<127.0.0.1> EXEC /bin/sh -c 'chmod u+x /Users/shertel/.ansible/tmp/ansible-tmp-1521813525.36-213064420310223/ /Users/shertel/.ansible/tmp/ansible-tmp-1521813525.36-213064420310223/s3_bucket.py && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '/tmp/ansible-test-coverage-hq_Wp7/coverage/injector.py /Users/shertel/.ansible/tmp/ansible-tmp-1521813525.36-213064420310223/s3_bucket.py && sleep 0'
<127.0.0.1> EXEC /bin/sh -c 'rm -f -r /Users/shertel/.ansible/tmp/ansible-tmp-1521813525.36-213064420310223/ > /dev/null 2>&1 && sleep 0'
changed: [localhost] => {
    "changed": true,
    "invocation": {
        "module_args": {
            "aws_access_key": "redacted",
            "aws_secret_key": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER",
            "ceph": false,
            "ec2_url": null,
            "force": false,
            "name": "ansible-test-shertel-osx-20178879-testbucket-ansible",
            "policy": null,
            "profile": null,
            "region": "us-east-1",
            "requester_pays": false,
            "s3_url": null,
            "security_token": "",
            "state": "present",
            "tags": null,
            "validate_certs": true,
            "versioning": null
        }
    },
    "name": "ansible-test-shertel-osx-20178879-testbucket-ansible",
    "policy": null,
    "requester_pays": false,
    "tags": {},
    "versioning": {
        "MfaDelete": "Disabled",
        "Versioning": "Disabled"
    }
}

TASK [s3_bucket : assert] ******************************************************************************************************
task path: /Users/shertel/Workspace/ansible/test/integration/targets/s3_bucket/tasks/main.yml:37
fatal: [localhost]: FAILED! => {
    "assertion": "not output.changed",
    "changed": false,
    "evaluated_to": false
}
```